File: Program's_Contributed_By_Contributors/C++ Programs/BinarySearch/bookallocation.cpp
/* C++ code for Book Allocation problem to find the minimum value of the maximum
number of pages*/
#include <bits/stdc++.h>

using namespace std;
/*function to check if it is possible to allocate the books such that the
maximum number of pages assigned to any student is numPages*/
bool isPossible(int pages[], int n, int m, int numPages) {
    int cntStudents = 1;
    int curSum = 0;
    for (int i = 0; i < n; i++) {
        if (pages[i] > numPages) {
            return false;
        }
        if (curSum + pages[i] > numPages) {
            /* Increment student count by '1'*/
            cntStudents += 1;
            /* assign current book to next student and update curSum */
            curSum = pages[i];
            /* If count of students becomes greater than given no. of students, return False*/
            if (cntStudents > m) {
                return false;
            }
        } else {
            /* Else assign this book to current student and update curSum */
            curSum += pages[i];
        }
    }
    return true;
}
int allocateBooks(int pages[], int n, int m) {
    /* If number student is more than number of books */
    if (n < m) {
        return -1;
    }
    /* Count total number of pages */
    int sum = 0;
    for (int i = 0; i < n; i++) {
        sum += pages[i];
    }
    /* Check for every possible value */
    for (int numPages = 1; numPages <= sum; numPages++) {
        if (isPossible(pages, n, m, numPages)) {
            return numPages;
        }
    }
    return -1;
}
int main() {
    int n = 4;
    int m = 2;
    int pages[] = {10, 20, 30, 40};
    cout << "The minimum value of the maximum number of pages in book allocation is: " << allocateBooks(pages, n, m) << endl;
}

File: Program's_Contributed_By_Contributors/Java_Programs/java/Book_Allocation.java
//Problem : https://www.geeksforgeeks.org/allocate-minimum-number-pages/

// Java program for optimal allocation of pages

public class Book_Allocation {
	// Utility method to check if current minimum value
	// is feasible or not.
	static boolean isPossible(int arr[], int n, int m,
							int curr_min)
	{
		int studentsRequired = 1;
		int curr_sum = 0;

		// iterate over all books
		for (int i = 0; i < n; i++) {
			curr_sum += arr[i];
			if (curr_sum > curr_min) {
				studentsRequired++; // increment student
									// count

				curr_sum = arr[i]; // update curr_sum
			}
		}

		return studentsRequired <= m;
	}

	// method to find minimum pages
	static int findPages(int arr[], int n, int m)
	{
		int sum = 0;

		// return -1 if no. of books is less than
		// no. of students
		if (n < m)
			return -1;

		// Count total number of pages
		for (int i = 0; i < n; i++)
			sum += arr[i];

		// initialize start as arr[n-1] pages(minimum answer
		// possible) and end as total pages(maximum answer
		// possible)
		int start = arr[n - 1], end = sum;
		int result = Integer.MAX_VALUE;

		// traverse until start <= end
		while (start <= end) {
			// check if it is possible to distribute
			// books by using mid is current minimum
			int mid = start + (end - start) / 2;
			if (isPossible(arr, n, m, mid)) {
				// update result to current distribution
				// as it's the best we have found till now.
				result = mid;

				// as we are finding minimum so,
				end = mid - 1;
			}

			else
				// if not possible, means pages should be
				// increased ,so update start = mid + 1
				start = mid + 1;
		}

		// at-last return minimum no. of pages
		return result;
	}

	// Driver Method
	public static void main(String[] args)
	{

		int arr[] = { 12, 34, 67,
					90 }; // Number of pages in books

		int m = 2; // No. of students

		System.out.println("Minimum number of pages = "
						+ findPages(arr, arr.length, m));
	}
}
